File: gnd/src/main/java/com/google/android/gnd/ui/home/AddFeatureDialogFragment.java
/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.gnd.ui.home;

import static com.google.android.gnd.util.ImmutableListCollector.toImmutableList;
import static java.util.Objects.requireNonNull;
import static java8.util.stream.StreamSupport.stream;

import android.app.Dialog;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.core.util.Consumer;
import androidx.fragment.app.FragmentManager;
import com.google.android.gnd.R;
import com.google.android.gnd.model.layer.Layer;
import com.google.android.gnd.ui.common.AbstractDialogFragment;
import com.google.common.collect.ImmutableList;
import dagger.hilt.android.AndroidEntryPoint;
import java8.util.Objects;
import javax.inject.Inject;
import timber.log.Timber;

@AndroidEntryPoint
public class AddFeatureDialogFragment extends AbstractDialogFragment {

  private static final String TAG = AddFeatureDialogFragment.class.getSimpleName();

  @Nullable private Consumer<Layer> layerConsumer;
  @Nullable private ImmutableList<Layer> layers;

  @Inject
  public AddFeatureDialogFragment() {}

  public void show(
      ImmutableList<Layer> layers, FragmentManager fragmentManager, Consumer<Layer> layerConsumer) {
    this.layers = layers;
    this.layerConsumer = layerConsumer;
    show(fragmentManager, TAG);
  }

  @NonNull
  @Override
  public Dialog onCreateDialog(Bundle savedInstanceState) {
    super.onCreateDialog(savedInstanceState);
    try {
      requireNonNull(layers);
      requireNonNull(layerConsumer);
      return createDialog(sortByName(layers), layerConsumer);
    } catch (RuntimeException e) {
      Timber.e(e);
      return fail(Objects.requireNonNullElse(e.getMessage(), "Unknown error"));
    }
  }

  private ImmutableList<Layer> sortByName(ImmutableList<Layer> layers) {
    return stream(layers)
        .sorted((l1, l2) -> l1.getName().compareTo(l2.getName()))
        .collect(toImmutableList());
  }

  private String[] getLayerNames(ImmutableList<Layer> layers) {
    return stream(layers).map(Layer::getName).toArray(String[]::new);
  }

  private Dialog createDialog(ImmutableList<Layer> layers, Consumer<Layer> layerConsumer) {
    // TODO: Add icons.
    return new AlertDialog.Builder(requireContext())
        .setTitle(R.string.add_feature_select_type_dialog_title)
        .setNegativeButton(R.string.cancel, (dialog, id) -> dismiss())
        .setItems(getLayerNames(layers), (dialog, index) -> layerConsumer.accept(layers.get(index)))
        .create();
  }

  @Override
  public void onDestroyView() {
    super.onDestroyView();
    layers = null;
    layerConsumer = null;
  }
}

File: gnd/src/main/java/com/google/android/gnd/ui/home/AddFeatureDialogFragment.kt
/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.android.gnd.ui.home

import android.app.Dialog
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject
import com.google.android.gnd.ui.common.AbstractDialogFragment
import android.os.Bundle
import com.google.android.gnd.R
import androidx.appcompat.app.AlertDialog
import androidx.core.util.Consumer
import androidx.fragment.app.FragmentManager
import com.google.android.gnd.model.layer.Job

@AndroidEntryPoint
class AddFeatureDialogFragment @Inject constructor() : AbstractDialogFragment() {

    private val fragmentTag = AddFeatureDialogFragment::class.java.simpleName
    private lateinit var jobConsumer: Consumer<Job>
    private lateinit var jobs: List<Job>

    fun show(
        jobs: List<Job>,
        fragmentManager: FragmentManager,
        jobConsumer: Consumer<Job>
    ) {
        this.jobs = jobs
        this.jobConsumer = jobConsumer
        show(fragmentManager, fragmentTag)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        super.onCreateDialog(savedInstanceState)
        return createDialog(sortByName(jobs), jobConsumer)
    }

    private fun sortByName(jobs: List<Job>): List<Job> = jobs.sortedBy { it.name }

    private fun getLayerNames(jobs: List<Job>): List<String> = jobs.map { it.name }

    private fun createDialog(
        jobs: List<Job>,
        jobConsumer: Consumer<Job>
    ): Dialog {
        // TODO: Add icons.
        return AlertDialog.Builder(requireContext())
            .setTitle(R.string.add_feature_select_type_dialog_title)
            .setNegativeButton(R.string.cancel) { _, _ -> dismiss() }
            .setItems(getLayerNames(jobs).toTypedArray()) { _, index -> jobConsumer.accept(jobs[index]) }
            .create()
    }
}
